docs/cpp.rst

```diff
@@ -0,0 +1,62 @@
+C++ domain customization
+========================
+
+.. confval:: cpp_strip_namespaces_from_signatures
+
+   :python:`list[str]` specifying namespaces to strip from signatures.  This
+   does not apply to the name of the symbol being defined by the signature, only
+   to parameter types, return types, default value expressions, etc.
+
+   For example, with the following in :file:`conf.py`:
+
+   .. code-block:: python
+
+      cpp_strip_namespaces_from_signatures = ["my_ns1", "my_ns2"]
+
+   .. rst-example:: C++ function definition with stripped namespaces
+
+      .. cpp:type:: my_ns1::A
+
+      .. cpp:type:: my_ns2::B
+
+      .. cpp:type:: my_ns3::C
+
+      .. cpp:function:: void my_ns1::MyFunction(my_ns1::A x, my_ns2::B y, my_ns3::C);
+
+   .. warning::
+
+      While intended for top-level namespaces, this strips any symbol with the
+      specified name, and does not take scope or other information into account.
+      For example, if you specify :python:`cpp_strip_namespaces_from_signatures
+      = ["abc"]`, then :cpp:`def::abc::xyz` will be converted to
+      :cpp:`def::xyz`.
+
+   .. rst-example:: Stripping of non-top-level symbols
+
+      .. cpp:function:: void MyFunction2(abc::my_ns1::A x);
+
+.. confval:: cpp_qualify_parameter_ids
+
+   :python:`bool` specifying whether function, template, and macro parameters
+   should be assigned fully-qualified id (for cross-linking purposes) of the
+   form ``<parent-id>-p-<param-name>`` based on the id of the parent
+   declaration.  Defaults to :python:`True`.
+
+   If set to :python:`False`, instead the shorter unqualified id
+   ``p-<param-name>`` is used.  This option should only be set to
+   :python:`False` if each C++ declaration is on a separate page.
+
+
+:cpp:`#include` directives in signatures
+----------------------------------------
+
+This theme extend the C and C++ domains to allow signatures to specify required
+:cpp:`#include` directives.
+
+.. rst-example:: Specifying :cpp:`#include` directives in signatures
+
+   .. cpp:function:: #include "my_header.h"
+                     #include "another_header.h"
+                     void foo(int param);
```

Comment:
The user shouldn't be surprised if clang-format re-orders `#include` statements when using that ext.

Reply:
Oops, I may be actually thinking of clang-tidy concerning the ordering, but clang-format also does some grouping and indenting with `#include`s.

Reply:
The format_signatures extension detects `#include` statements and ignores them in part because they are not normal signatures and also because as you mention reordering would be a problem (due to how it works, it cannot support reordering anyway).

Reply:
It would be cool if std libs like `cstdio` could be cross-referenced from include statements...

Reply:
Yeah, perhaps in a later commit --- would require some thought to decide how to cross link header files in a general way.